### ISSUE TITLE
Potential memory leak in CRehldsPlatformHolder

### ISSUE THREAD

**Issue by Sh1ft0x0EF** [dreamstalker/rehlds#377]
Current impl of CRehldsPlatformHolder class creates a new default impl of IRehldsPlatform interface in it's get method if it's m_Platform pointer is currently null:

https://github.com/dreamstalker/rehlds/blob/master/rehlds/rehlds/platform.cpp#L7

The set method allows to set a new impl of m_Platform without any check:

https://github.com/dreamstalker/rehlds/blob/master/rehlds/rehlds/platform.cpp#L14

so this code will occur in a memory leak:

```cpp
IReHLDSPlatform *pPlatform = CRehldsPlatformHolder::get(); // m_Platform points at null and default impl will be created
CRehldsPlatformHolder::set(nullptr); // m_Platform is now points at null again which makes default instantiated impl inaccessible
```

Proposed fix:
https://github.com/Sh1ft0x0EF/rehlds/commit/593caec5d34d42e67c235aa808232dbaa52154a6

**Comment by theAsmodai** [comment]
Or we can delete m_Platform when the setter used with nullptr.

**Comment by Sh1ft0x0EF** [comment]
If it's guaranteed that passed impl of IReHLDSPlatform in the set method is using the same crt (i.e. didn't came from any different dll from 3rd party) then there wouldn't be any problem to just delete it

**Comment by LevShisterov** [comment]
Actually, `set` is called only once in DllMain, so I see no reason there will be a leak.
But if imagination is applied, then you always can shoot in the leg with C++, for ex: yes, delete `m_Platform` in `set`, but if it was stored previously from the call to `get`, then it will crash on deleted object later, etc...

**Comment by Sh1ft0x0EF** [comment]
Even if you store a pointer you always should to validate it before use
The Service Locator pattern is designed the way you access the pointer through it, not store it

Alternative solution to save some CPU cycles when calling the get method:

```cpp
class CRehldsPlatformHolder
{
private:
 	static IReHLDSPlatform* m_Platform;

 	static CSimplePlatform DefaultPlatform;
public:
	static void init(){m_Platform = &DefaultPlatform;}
	static IReHLDSPlatform& get(){return *m_Platform;}
	static void set(IReHLDSPlatform* p)
	{
		if (!p)
		{
			m_Platform = &DefaultPlatform;
			return;
		};

		m_Platform = p;
	};
};

**Comment by LevShisterov** [comment]
There is no way to validate a pointer to deleted object. At least I dunno about direct ones.
Simplifying getter is a good idea.

**Comment by IgnacioFDM** [comment]
Indeed there is no way to validate pointers. There used to be some windows specific function some time ago, which was completely broken and has been dropped, and even then simply told you if it was "safe" to access, however the object could have been deleted, and a new unrelated object reside on the same address, therefore the pointer still being invalid.

**Comment by WPMGPRoSToTeMa** [comment]
We can use something like `unique_ptr` to simplify this.

**Comment by Sh1ft0x0EF** [comment]
shared_ptr has a double size in mem consumption and a double size reduction in speed because of ref counting mech that the unique_ptr

**Comment by WPMGPRoSToTeMa** [comment]
Also idk why we are using static class `CRehldsPlatformHolder` when we can just use global variable `unique_ptr<IReHLDSPlatform> g_RehldsPlatform`.

**Comment by Sh1ft0x0EF** [comment]
@WPMGPRoSToTeMa  It's useless to declare unique_ptr globally because it's used to automatically delete data allocated on the heap (instead of manually write a delete for it which is sometimes forgotten by devs and results in mem leaks).
IMHO better to use

```cpp
IRehldsPlatformHolder *gpPlatform;
```

or something like that then

What is the purpose of CRehldsPlatformHolder?
I saw some places where it was used only for Windows compilation and Unix was using POSIX directly
Maybe we should restore the direct usage of POSIX instead of platform holder where it's possible?

**Comment by WPMGPRoSToTeMa** [comment]
>It's useless to declare unique_ptr globally because it's used to automatically delete data allocated on the heap

Why?
>which is sometimes forgotten by devs and results in mem leaks

Here you write about it benefit, so why we shouldn't use it?

**Comment by Sh1ft0x0EF** [comment]
@WPMGPRoSToTeMa 
https://stackoverflow.com/questions/34534927/using-smart-pointers-as-global-variables
https://stackoverflow.com/questions/19867360/how-to-instantiate-a-global-smart-pointer-variable

If we declare it globally it's lifetime will be equal to app lifetime
I've proposed to use it for internal default platform impl CSimplePlatform, not for platform holder
Do you want to use it globally to be able to reset it to other impl? I mean unique_ptr could be used as holder itself so we could reset it's internal pointer and use some other implementation

**Comment by IgnacioFDM** [comment]
Because a global variable never goes out of scope, therefore it's destructor is never called.

**Comment by Sh1ft0x0EF** [comment]
@IgnacioFDM , all global variables will be released at program shutdown

**Comment by Sh1ft0x0EF** [comment]
@WPMGPRoSToTeMa 
> Also idk why we are using static class CRehldsPlatformHolder instead of global variable unique_ptr<IReHLDSPlatform> g_RehldsPlatform.

CRehldsPlatformHolder has a small and simple interface - only get and set methods
How about using the smart pointer for internal IReHLDSPlatform interface of the platform holder?

**Comment by IgnacioFDM** [comment]
@Sh1ft0x0EF Yes you're right, I forgot about that. It can be very important in terms of properly releasing resources.

**Comment by WPMGPRoSToTeMa** [comment]
> How about using the smart pointer for internal IReHLDSPlatform interface of the platform holder?

Yeah its a good idea.

**Comment by In-line** [comment]
We should also check libstdc++ version requirement for `std::unique_ptr`.

**Comment by Phantomsq** [comment]
So, is this bug is there now?

**Comment by Sh1ft0x0EF** [comment]
@Phantomsq 
Yep, it's still there
I think I'll fix it on this weekend

**Comment by LevShisterov** [comment]
No bugs here, because CRehldsPlatformHolder::set is only used in testing suite. And even there it is executed only once on engine loading.
